libs/datamodel/core/src/transform/ast_to_dml/db/types.rs

```diff
@@ -0,0 +1,88 @@
+use super::ScalarFieldType;
+use crate::{
+    ast::{FieldId, SchemaAst, TopId},
+    diagnostics::{DatamodelError, Diagnostics},
+};
+use itertools::Itertools;
+use std::collections::{BTreeMap, BTreeSet, HashMap};
+
+type AnnotationId = u32;
+
+#[derive(Debug, Default)]
+pub(super) struct Types {
+    pub(super) type_aliases: BTreeMap<TopId, ScalarFieldType>,
+    pub(super) scalar_fields: BTreeMap<(TopId, FieldId), ScalarFieldType>,
+    annotations: Vec<ModelAnnotation>,
+    // Storage for annotation fields, i.e. the fields referenced in `@(@)index`,
+    // `@(@)unique` and `@(@)id`. The type should be understood as (model_id,
+    // annotation_id, order_key_in_annotation, field_id)
+    annotation_fields: BTreeSet<(TopId, AnnotationId, u16, FieldId)>,
+}
+
+impl Types {
+    /// Detect self-referencing type aliases (possibly indirectly).
+    pub(super) fn detect_alias_cycles(&self, ast: &SchemaAst, diagnostics: &mut Diagnostics) {
```

Comment:
did we have this in the old impl?

Reply:
BTW. this part is quite hard to follow. It's useful for sure, also for referential actions later in the pipeline.

But, I'd reconsider the code a bit before merging here. You can even make it slower, but it's better when it's easy to follow.

Reply:
>  did we have this in the old impl?

Yep, it's still there. It bails at the first error though, whereas this one will detect all cycles in the schema.

>  this part is quite hard to follow.

Yes maybe we should have a call about it. At the minimum, we can write a comment with the purpose of this looping. I think implementation-wise, it's either this loop or recursion, but I think the recursive implementation would also be hard to follow.

Reply:
I wrote more docs, please check them out.